File: others/gamecore/mapcore_py/editor/tools.py
"""地形雕刻工具：筆刷、山脊、裂谷。"""
from __future__ import annotations
import math
import random as _random
from .state import EditorState
from .hex_layout import hex_disk, hex_distance


def apply_brush(state: EditorState, q: int, r: int, delta: float) -> None:
    """高斯衰減筆刷：中心 delta 最強，邊緣趨近 0。"""
    for nq, nr in hex_disk(q, r, state.brush_size):
        if not state.in_bounds(nq, nr):
            continue
        dist = hex_distance(q, r, nq, nr)
        weight = _gaussian(dist, state.brush_size)
        state.set_h(nq, nr, state.get_h(nq, nr) + delta * weight)


# ── Radial stamp brush (editor interactive) ──────────────────────────────────

def apply_ridge_stamp(state: EditorState, q: int, r: int) -> None:
    """以滑鼠為中心的徑向山脊筆刷（非線性衰減，可加雜亂度與放射線）。"""
    _apply_radial_stamp(state, q, r, +1)


def apply_rift_stamp(state: EditorState, q: int, r: int) -> None:
    """以滑鼠為中心的徑向裂谷筆刷（非線性衰減，可加雜亂度與放射線）。"""
    _apply_radial_stamp(state, q, r, -1)


def _tile_hash(q: int, r: int) -> float:
    """每格穩定的偽隨機值 0..1，不依賴 random 模組。"""
    v = (q * 1664525 + r * 1013904223) & 0xFFFFFFFF
    v ^= v >> 16
    v  = (v * 0x45d9f3b) & 0xFFFFFFFF
    v ^= v >> 16
    return (v & 0xFFFF) / 65535.0


def _apply_radial_stamp(state: EditorState, q: int, r: int, sign: float) -> None:
    radius   = max(1, state.brush_size)
    strength = state.brush_strength * 2.5
    chaos    = state.brush_chaos
    falloff  = state.brush_falloff
    invert   = state.brush_spokes_invert

    # ── 決定本次 stamp 的放射線數量與旋轉偏移 ────────────────────
    if state.brush_spokes_rand:
        lo = min(state.brush_spokes_min, state.brush_spokes_max)
        hi = max(state.brush_spokes_min, state.brush_spokes_max)
        n_spokes     = _random.randint(lo, hi)
        spoke_offset = _random.uniform(0.0, math.pi * 2.0) if n_spokes > 0 else 0.0
    else:
        n_spokes     = state.brush_spokes
        spoke_offset = 0.0

    # 搜尋半徑：chaos 會擴展邊界；spokes 延伸到 2× 基礎半徑
    max_reach = radius * (1.0 + 0.6 * chaos)
    if n_spokes > 0:
        max_reach = max(max_reach, radius * 2.0)
    search_r = int(math.ceil(max_reach)) + 1

    for nq, nr in hex_disk(q, r, search_r):
        if not state.in_bounds(nq, nr):
            continue
        dist = hex_distance(q, r, nq, nr)

        # ── 徑向分量（非線性衰減 + 雜亂度）───────────────────────
        if chaos > 0.0:
            noise_val  = _tile_hash(nq, nr)
            eff_radius = radius * (1.0 + chaos * (noise_val * 2.0 - 1.0) * 0.55)
            eff_radius = max(0.5, eff_radius)
        else:
            eff_radius = float(radius)
        t           = max(0.0, 1.0 - dist / eff_radius)
        base_weight = t ** falloff if t > 0.0 else 0.0

        # ── 放射線分量（可反轉方向）──────────────────────────────
        spoke_w = 0.0
        if n_spokes > 0 and dist > 0:
            dq    = nq - q
            dr    = nr - r
            # axial 偏移 → 近似笛卡爾
            cx_f  = dq + 0.5 * dr
            cy_f  = dr * 0.8660254
            angle     = math.atan2(cy_f, cx_f) - spoke_offset
            spoke_inc = math.pi * 2.0 / n_spokes
            idx       = round(angle / spoke_inc)
            nearest   = idx * spoke_inc
            diff      = abs(((angle - nearest) + math.pi) % (math.pi * 2.0) - math.pi)
            half_w    = spoke_inc * 0.35
            if diff < half_w:
                spoke_cos = math.cos(diff / half_w * math.pi * 0.5)
                spoke_rad = max(0.0, 1.0 - dist / (radius * 2.0))
                spoke_w   = spoke_cos ** 2 * spoke_rad

        # ── 合併：同向取 max，反向做差值 ─────────────────────────
        if invert:
            # 放射線和主工具方向相反：基礎推高、放射線壓低（或反之）
            effective = base_weight - spoke_w
        else:
            effective = max(base_weight, spoke_w)

        if abs(effective) < 1e-6:
            continue
        state.set_h(nq, nr, state.get_h(nq, nr) + sign * strength * effective)


def toggle_water_source(state: EditorState, q: int, r: int) -> None:
    pt = (q, r)
    if pt in state.water_sources:
        state.water_sources.remove(pt)
    else:
        state.water_sources.append(pt)


def _gaussian(dist: float, radius: float) -> float:
    if radius <= 0:
        return 1.0
    sigma = radius / 2.2
    return math.exp(-0.5 * (dist / sigma) ** 2)

File: others/gamecore/mapcore_cpp_square/viewer/tools.cpp
#include "tools.hpp"

#include "grid_layout.hpp"

#include <algorithm>
#include <cmath>
#include <cstdint>

namespace viewer {

int ToolRng::uniform_int(int lo, int hi) {
    if (hi < lo) std::swap(lo, hi);
    std::uniform_int_distribution<int> d(lo, hi);
    return d(engine_);
}

float ToolRng::uniform_float(float lo, float hi) {
    if (hi < lo) std::swap(lo, hi);
    std::uniform_real_distribution<float> d(lo, hi);
    return d(engine_);
}

// ── 高斯筆刷 ──────────────────────────────────────────────────────────

namespace {

[[nodiscard]] float gaussian(float dist, float radius) noexcept {
    if (radius <= 0.0f) return 1.0f;
    const float sigma = radius / 2.2f;
    const float n = dist / sigma;
    return std::exp(-0.5f * n * n);
}

} // anonymous

void apply_brush(EditorState& s, int x, int y, float delta) {
    const int radius = std::max(1, s.brush_size);
    // 走方形 disk，但用 Euclidean 距離當衰減：圓形視覺感、不浪費 ring 形狀
    for (const GridCoord& h : grid_disk(x, y, radius)) {
        if (!s.in_bounds(h.x, h.y)) continue;
        const float dxf = static_cast<float>(h.x - x);
        const float dyf = static_cast<float>(h.y - y);
        const float dist = std::sqrt(dxf * dxf + dyf * dyf);
        if (dist > static_cast<float>(radius) + 0.5f) continue;
        const float w = gaussian(dist, static_cast<float>(radius));
        s.add_h(h.x, h.y, delta * w);
    }
}

// ── Radial stamp（ridge / rift） ──────────────────────────────────────

namespace {

// 對齊 tools.py:_tile_hash —— 每格穩定的偽隨機值 [0, 1)。
[[nodiscard]] float tile_hash(int x, int y) noexcept {
    uint32_t v = static_cast<uint32_t>(x) * 1664525u
               + static_cast<uint32_t>(y) * 1013904223u;
    v ^= v >> 16;
    v *= 0x45d9f3bu;
    v ^= v >> 16;
    return static_cast<float>(v & 0xFFFFu) / 65535.0f;
}

constexpr float kPi  = 3.14159265358979323846f;
constexpr float kTau = 6.28318530717958647692f;

void apply_radial_stamp(EditorState& s, int cx, int cy, float sign, ToolRng& rng) {
    const int   radius   = std::max(1, s.brush_size);
    const float strength = s.brush_strength * 2.5f;
    const float chaos    = s.brush_chaos;
    const float falloff  = s.brush_falloff;
    const bool  invert   = s.brush_spokes_invert;

    int   n_spokes     = s.brush_spokes;
    float spoke_offset = 0.0f;
    if (s.brush_spokes_rand) {
        const int lo = std::min(s.brush_spokes_min, s.brush_spokes_max);
        const int hi = std::max(s.brush_spokes_min, s.brush_spokes_max);
        n_spokes = rng.uniform_int(lo, hi);
        if (n_spokes > 0) spoke_offset = rng.uniform_float(0.0f, kTau);
    }

    float max_reach = static_cast<float>(radius) * (1.0f + 0.6f * chaos);
    if (n_spokes > 0) max_reach = std::max(max_reach, static_cast<float>(radius) * 2.0f);
    const int search_r = static_cast<int>(std::ceil(max_reach)) + 1;

    const float spoke_inc = (n_spokes > 0) ? (kTau / static_cast<float>(n_spokes)) : 0.0f;
    const float half_w    = spoke_inc * 0.35f;

    for (const GridCoord& h : grid_disk(cx, cy, search_r)) {
        if (!s.in_bounds(h.x, h.y)) continue;
        const float dxf = static_cast<float>(h.x - cx);
        const float dyf = static_cast<float>(h.y - cy);
        const float dist = std::sqrt(dxf * dxf + dyf * dyf);

        // 徑向分量：非線性衰減 + 雜亂度
        float eff_radius;
        if (chaos > 0.0f) {
            const float noise_val = tile_hash(h.x, h.y);
            eff_radius = static_cast<float>(radius) * (1.0f + chaos * (noise_val * 2.0f - 1.0f) * 0.55f);
            if (eff_radius < 0.5f) eff_radius = 0.5f;
        } else {
            eff_radius = static_cast<float>(radius);
        }
        const float t = std::max(0.0f, 1.0f - dist / eff_radius);
        const float base_weight = (t > 0.0f) ? std::pow(t, falloff) : 0.0f;

        // 放射線分量
        float spoke_w = 0.0f;
        if (n_spokes > 0 && dist > 0.0f) {
            const float angle = std::atan2(dyf, dxf) - spoke_offset;
            const float idx   = std::round(angle / spoke_inc);
            const float nearest = idx * spoke_inc;
            float raw_diff = angle - nearest;
            float diff = std::fmod(raw_diff + kPi, kTau);
            if (diff < 0.0f) diff += kTau;
            diff = std::abs(diff - kPi);
            if (diff < half_w) {
                const float spoke_cos = std::cos(diff / half_w * kPi * 0.5f);
                const float spoke_rad = std::max(0.0f, 1.0f - dist / (static_cast<float>(radius) * 2.0f));
                spoke_w = spoke_cos * spoke_cos * spoke_rad;
            }
        }

        const float effective = invert ? (base_weight - spoke_w)
                                       : std::max(base_weight, spoke_w);
        if (std::abs(effective) < 1e-6f) continue;
        s.add_h(h.x, h.y, sign * strength * effective);
    }
}

} // anonymous

void apply_ridge_stamp(EditorState& s, int x, int y, ToolRng& rng) {
    apply_radial_stamp(s, x, y, +1.0f, rng);
}

void apply_rift_stamp(EditorState& s, int x, int y, ToolRng& rng) {
    apply_radial_stamp(s, x, y, -1.0f, rng);
}

// ── Water source toggle ──────────────────────────────────────────────

void toggle_water_source(EditorState& s, int x, int y) {
    const GridCoord pt{x, y};
    auto it = std::find(s.water_sources.begin(), s.water_sources.end(), pt);
    if (it == s.water_sources.end()) s.water_sources.push_back(pt);
    else                              s.water_sources.erase(it);
}

} // namespace viewer
